packages/deployer/src/deployer.ts

```diff
@@ -21,458 +21,93 @@
-            createdSolutionId => {
-              /* istanbul ignore else */
-              if (deployOptions.progressCallback) {
-                deployOptions.progressCallback(100); // we're done
-              }
-              resolve(createdSolutionId);
-            },
-            error => {
-              // Error deploying solution
-              /* istanbul ignore else */
-              if (deployOptions.progressCallback) {
-                deployOptions.progressCallback(1);
-              }
-              reject(error);
-            }
-          );
+  // if we are not passed the maybeModel, reject
+  if (!maybeModel) {
+    return Promise.reject(common.fail("The Solution Template id is missing"));
+  }
+  let deployOptions: common.IDeploySolutionOptions = options || {};
+  /* istanbul ignore else */
+  if (deployOptions.progressCallback) {
+    deployOptions.progressCallback(1); // let the caller know that we've started
+  }
+  // deal with maybe getting an item or an id
+  return _getSolutionTemplateItem(maybeModel, authentication)
+    .then(model => {
+      if (!isSolutionTemplateItem(model.item)) {
+        return Promise.reject(
+          common.fail(`${model.item.id} is not a Solution Template`)
+        );
+      } else {
+        // fetch the metadata and pass the item & data forward
+        return Promise.all([
+          Promise.resolve(model.item),
+          Promise.resolve(model.data),
+          common.getItemMetadataAsFile(model.item.id, authentication)
+        ]);
+      }
+    })
+    .then(responses => {
+      // extract responses
+      const [itemBase, itemData, itemMetadata] = responses;
+      // sanitize all the things
+      const sanitizer = new common.Sanitizer();
+      const item = common.sanitizeJSONAndReportChanges(itemBase, sanitizer);
+      // TODO: we should delegate data sanization to the type-specific modules
+      const data = common.sanitizeJSONAndReportChanges(itemData, sanitizer);
+      // apply item props to deployOptions
+      deployOptions = _updateDeployOptions(deployOptions, item, authentication);
+      // Clone before mutating? This was messing me up in some testing...
+      common.deleteItemProps(item);
+
+      return _deploySolutionFromTemplate(
+        item.id,
+        item,
+        data,
+        itemMetadata,
+        authentication,
+        deployOptions
+      );
+    })
+    .then(
+      createdSolutionId => {
+        /* istanbul ignore else */
+        if (deployOptions.progressCallback) {
+          deployOptions.progressCallback(100); // we're done
         }
+        return createdSolutionId;
       },
       error => {
-        // Error fetching solution
+        // Error deploying solution
         /* istanbul ignore else */
         if (deployOptions.progressCallback) {
           deployOptions.progressCallback(1);
         }
-        reject(error);
+        return Promise.reject(error);
       }
-    );
-  });
+    )
+    .catch(ex => {
+      throw ex;
```

Comment:
Please change to Promise.reject--the several current clients of the library only have to support Promise returns

Reply:
That will return a rejected promise :)